### DISCUSSION TITLE
Understrap install with composer and WPackagist

### DISCUSSION THREAD

**Discussion by Guillaume-Tnk** [holger1411/understrap#1301] [Q&A]
Hello, 
does anyone know how to install understrap theme with composer ? 
I was able to install understrap through the composer.json file
```
"require": {
        "wpackagist-theme/understrap": "0.9.6"
    },
 ```
 but the child theme is not available [here](https://wpackagist.org/search?q=understrap&type=any&search=)
 is there a plan to add the child theme free version to wpackagist or should I proceed another way?

**Comment by 0dp** [comment]
If an updated version of US is available on packagist then it's only right that the child theme is too.

Thanks for bringing attention to this.

Is installing the theme via composer documented or do we need a writeup from @understrap/documentation-team

**Reply by Guillaume-Tnk** [reply]
Thank you @0dp for your reply, in the meantime I'm trying this
```
"repositories": [{
            "type": "composer",
            "url": "https://wpackagist.org",
            "only": ["wpackagist-plugin/*", "wpackagist-theme/*"]
        },
        {
            "type": "package",
            "package": {
                "name": "understrap/understrap-child",
                "type": "wordpress-theme",
                "version": "0.5.1",
                "source": {
                    "type": "git",
                    "url": "https://github.com/understrap/understrap-child.git",
                    "reference": "main"
                }
            }
        }
    ],
    ```
    ```
"require": {
        "wpackagist-theme/understrap": "0.9.6",
        "understrap/understrap-child": "0.5.1"
    },
```
based on reading [this](https://easyengine.io/tutorials/composer-wordpress/manage-themes-plugins/)
but it's not installing, I must be missing something ...

any advice?

**Reply by danemorgan** [reply]
Is the version on packagist up to date? Do they automatically follow our releases, or do we have to push updates to them?

**Reply by danemorgan** [reply]
If we're going to support composer installs, it should probably be documented in the readme.md.

**Comment by Guillaume-Tnk** [comment]
Ok, found a working solution with the same [link](https://easyengine.io/tutorials/composer-wordpress/manage-themes-plugins/) using the zip option
```
"repositories": [{
            "type": "composer",
            "url": "https://wpackagist.org",
            "only": ["wpackagist-plugin/*", "wpackagist-theme/*"]
        },
        {
            "type": "package",
            "package": {
                "name": "understrap/understrap-child",
                "type": "wordpress-theme",
                "version": "0.5.1",
                "dist": {
                    "type": "zip",
                    "url": "https://github.com/understrap/understrap-child/archive/refs/heads/main.zip"
                }
            }
        }
    ],
    "require": {
        ...
        "wpackagist-theme/understrap": "0.9.6",
        "understrap/understrap-child": "0.5.1"
    },

```
    Is it possible to notify on this post if/when you make the child theme available on wpackagist , please?

**Reply by 0dp** [reply]
Great you found a workaround. 👍

If i were you, I'd raise the issue on the child-theme.
Then you'll be sure to get notified when the child-theme is registered with packagist.

**Reply by howarddc** [reply]
Thanks for bringing this up - we'll work on a permanent fix in our upcoming releases.

**Reply by bacoords** [reply]
Just one question - if the child theme is installed via composer, will any updates we push overwrite your custom code? 

I'm not super experienced with composer, but just want to make sure that anyone installing it that way knows if there are any additional steps to make sure they don't get updated pulled automatically.

**Reply by danemorgan** [reply]
Wouldn't you have to run an update or install command through composer?

**Reply by danemorgan** [reply]
[Composer basic usage](https://getcomposer.org/doc/01-basic-usage.md)

> #Updating dependencies to their latest versions#
>As mentioned above, the composer.lock file prevents you from automatically getting the latest versions of your dependencies. To update to the latest versions, use the update command. This will fetch the latest matching versions (according to your composer.json file) and update the lock file with the new versions.
>
>`php composer.phar update`
>Note: Composer will display a Warning when executing an install command if the composer.lock has not been updated since changes were made to the composer.json that might affect dependency resolution.
>
>If you only want to install, upgrade or remove one dependency, you can explicitly list it as an argument:
>
>`php composer.phar update monolog/monolog [...]`

**Reply by danemorgan** [reply]
It seems like there be dragons, but that it is also manageable if documented correctly.

**Reply by Guillaume-Tnk** [reply]
@bacoords If your question was for me, I'm not advanced either with composer but as far as I  understand it will download the version of the child theme specified in the composer.json unless you specify that you always want the latest version.

**Comment by danemorgan** [comment]
Okay, we have to include the repository in our composer.json and then log in to packagist and post the URL.

~~@howarddc did we get a login for packagist from @holger1411?~~

It looks like we don't need a specific login. I think I can just submit it logged in as me.

**Reply by UnderstrapFramework** [reply]
Agreed, I don't have anything on file but we can create whatever is needed.

**Reply by IanDelMar** [reply]
Check out https://github.com/semantic-release!

Also have a look at the following packages:
```json
    "@google/semantic-release-replace-plugin": "^1.0.2",
    "@semantic-release/changelog": "^5.0.1",
    "@semantic-release/commit-analyzer": "^8.0.1",
    "@semantic-release/git": "^9.0.0",
    "@semantic-release/npm": "^7.1.3",
    "@semantic-release/release-notes-generator": "^9.0.3",
    "conventional-changelog-conventionalcommits": "^4.6.0",
    "semantic-release": "^17.4.4",
```
and also @semantic-release/github. You can create releases for Github and npm automatically.

Here's a .releaserc file _not_ utilizing npm and github releases
```json
{
	"branches": [
		"main"
	],
	"plugins": [
		"@semantic-release/commit-analyzer",
		"@semantic-release/release-notes-generator",
		[
			"@semantic-release/changelog",
	  		{
				"changelogFile": "./CHANGELOG.md"
	  		}
		],
		[
			"@google/semantic-release-replace-plugin",
			{
				"replacements": [
					{
						files: [
							"*.php",
							"build/**/*.js",
							"build/**/*.cjs",
							"inc/**/*.php"
						],
						from: "x\\.x\\.x",
						to: "${nextRelease.version}",
						ignore: [
							"vendor/**",
							"node_modules/**",
						],
						encoding: "utf-8",
						disableGlobs: false,
						countMatches: false,
						allowEmptyPaths: false,
					},
					{
						files: "style.css",
						from: "(Version:\\s*)\\d{1,2}\\.\\d{1,2}\\.\\d{1,2}",
						to: "$1${nextRelease.version}",
						encoding: "utf-8",
						disableGlobs: true,
						countMatches: false,
						allowEmptyPaths: false,
					},
					{
						files: "assets/js/theme*.js",
						from: "(Punk Buddha v)\\d{1,2}\\.\\d{1,2}\\.\\d{1,2}",
						to: "$1${nextRelease.version}",
						encoding: "utf-8",
						disableGlobs: false,
						countMatches: false,
						allowEmptyPaths: false,
					}
				]
			}
		],
		[
			"@semantic-release/npm",
			{
				"npmPublish": false,
  			}
		],
		[
			"@semantic-release/git",
			{
				"assets": [
					"assets/js/theme*.js",
					"build/**/*.js",
					"build/**/*.cjs",
					"inc/**/*.php",
					"*.php",
					"./CHANGELOG.md",
					"./package.json",
					"./package-lock.json",
					"./style.css"
				],
				"message": "chore(release): ${nextRelease.version} \n\n${nextRelease.notes}"
			}
		]
	],
	"preset": "conventionalcommits"
}
```

When thinking of automation, here's a script I use during build and release to update the version number in style.css.
```js
#!/usr/bin/env node

/**
 * Script to write style.css.
 *
 * @since 1.0.0
 */

import chalk from 'chalk'
import fs from 'fs'
import theme from './theme.config.js'
import path from 'path'

// Sanitize optional properties.
theme.uri = theme.uri && theme.uri.trim() ? theme.uri : null
theme.authorUri = theme.authorUri && theme.authorUri.trim() ? theme.authorUri : null
theme.template = theme.template && theme.template.trim() ? theme.template : null
theme.tags = theme.tags && theme.tags.trim() ? theme.tags : null
theme.domainPath = theme.domainPath && theme.domainPath.trim() ? theme.domainPath : null

let banner = `/*!
	Theme Name:        ${theme.name}
	${theme.uri ? 'Theme URI:         '+theme.uri : ''}
	Description:       ${theme.description}
	Author:            ${theme.author}
	${theme.authorUri ? 'Author URI:        '+theme.authorUri : ''}
	${theme.template ? 'Template:          '+theme.template : ''}
	${theme.tags ? 'Tags:              '+theme.tags : ''}
	Version:           ${theme.version}
	Requires at least: ${theme.requiresWp}
	Tested up to:      ${theme.testedUp}
	Requires PHP:      ${theme.requiresPhp}
	License:           ${theme.license}
	License URI:       ${theme.licenseUri}
	Text Domain:       ${theme.textDomain}
	${theme.domainPath ? 'Domain Path:       '+theme.domainPath : ''}
	*/`.replace(/(^\s*[\r\n]|\t)/gm, '')+'\n'
const file = './style.css'

console.log('Writing style.css...')

try {
	fs.writeFileSync(file, banner)
	console.log(banner);
	console.log(chalk.green('Banner written to ' + path.resolve(file) ))
} catch(err) {
	if (err) console.log(err);
}
```
calling it with this script `css-style": "node ./build/style-css.js",` in my package.json.

I use the `@google/semantic-release-replace-plugin` for inline changelogs. When adding code I use `@since x.x.x` in the doc block and the package then replaces this with the version number of the next release.

Sticking to a commit message convention is of course much harder for a public repo.

**Comment by danemorgan** [comment]
Let me advocate the devil here and ask if we have Node and if we move to Webpack, do we really need to also be doing composer and WPackagist?

**Reply by pattonwebz** [reply]
The tooling the theme itself uses to build isn't related to the ask in this issue.

The issue is about the way theme is distributed.

**Reply by danemorgan** [reply]
Right. Sorry. I keep mixing these things.

**Reply by pattonwebz** [reply]
No worries! Happens to us all lol
